Single indirect

kstowe opened the pull request:
I tested this on a file that was larger than the 10 direct blocks and it works.

kstowe committed:
Start single indirect

kstowe committed:
Finish indirectBlock

kstowe committed:
Finish single-indirect

kstowe commented on IndexNode.h:
```diff
@@ -11,7 +11,7 @@ class IndexNode
 		const static int INDEX_NODE_SIZE = 64 ;
 		const static int MAX_DIRECT_BLOCKS = 10 ;
-		const static int MAX_FILE_BLOCKS = MAX_DIRECT_BLOCKS ;
+		const static int MAX_FILE_BLOCKS =  MAX_DIRECT_BLOCKS;// + blockSize / 4
```
I hardcoded this when I tested, but MAX_FILE_BLOCKS should be proportional to the blockSize. However, I'm not exactly sure how to access the blockSize from the SuperBlock at this point. @elliott-beach Any ideas?

elliott-beach replied:
I would just remove that constant and call `fileSystem.getBlockSize` in fileDescriptor, which is where the constant is used.

elliott-beach replied:
Also, just so were on the same page, a test script I would use is 
```
./mkfs filesys.dat 64 64
python -c 'print("x" * 15 * 64)' | ./tee /bigfile
```

elliott-beach commented:
Phooey. I was working on this but you finished before me. I should have communicated more or spent less time eating nutella bread. 

I think you could save yourself the burden of the bitshift logic by just casting the `ptrBlock` array from `char*` to `int*` and accessing normally.

elliott-beach committed:
replace MAX_BLOCK_SIZE const with function check

elliott-beach commented:
And I utilized GitHub's ability to make the suggested change, hope you don't mind.

elliott-beach commented on IndexNode.cc:
```diff
@@ -127,17 +143,56 @@ int IndexNode::getBlockAddress(int block)
+		newBlock[i+3] = (unsigned char)(NOT_A_BLOCK);
+	    }
+	    // Fill in the address of the new block
+	    newBlock[(block-10)*4] = (unsigned char)(address >> 24);
```
This is missing a check on the value of `block`, methinks.

elliott-beach committed:
1. check block# when creating first block 2. simplify logic

I think using a cast from int to char* is a lot better than using
manual bit shifting, and it is definitely fewer lines of code.

elliott-beach commented on IndexNode.cc:
```diff
@@ -127,17 +138,40 @@ int IndexNode::getBlockAddress(int block)
+			// Pick an address for the indirectBlock
+			indirectBlock = fileSystem->allocateBlock();
+			if(indirectBlock < 0 ) {
+				return; // Maybe we should exit?
```
What we really want to do is make `Kernel::write` or `read` return -1 in this case. To be thorough, the best solution is to change the function's signature to return an exit status. Then the caller checks for an error.

elliott-beach replied:
That's hard tho

kstowe replied:
Yeah that's true. Whoever wrote these functions originally just decided to skip the error checking here.

elliott-beach commented on IndexNode.cc:
```diff
@@ -146,51 +141,35 @@ int IndexNode::getBlockAddress(int block)
-	    // Initialize all the ptrs in indirectBlock to NOT_A_BLACK
-	    char newBlock[blockSize];
-	    for(int i = 0; i < blockSize; i+=4) {
-		newBlock[i] = (unsigned char)(NOT_A_BLOCK >> 24);
```
I "simplified" this code. The difference probably doesn't matter much, so I hope you don't mind.

kstowe replied:
No it looks good. Your way is much easier to read

kstowe commented on FileDescriptor.cc:
```diff
@@ -125,7 +126,7 @@ int FileDescriptor::readBlock(short relativeBlockNumber)
 int FileDescriptor::writeBlock(short relativeBlockNumber) 
 {
-	if(relativeBlockNumber >= IndexNode::MAX_FILE_BLOCKS)
+	if(relativeBlockNumber >= fileSystem->getBlockSize() + IndexNode::MAX_DIRECT_BLOCKS)
```
I think that this needs to be `fileSystem->getBlockSize()/4` because each address is 4 bytes.
But this is a good solution.

elliott-beach replied:
oh yes

kstowe commented on IndexNode.cc:
```diff
@@ -109,13 +111,22 @@ int IndexNode::getSize()
+	else if(indirectBlock == NOT_A_BLOCK) {
+	    return indirectBlock;
+	}
+	else if(block < fileSystem->getBlockSize() + MAX_DIRECT_BLOCKS) {
```
Same here with the blockSize()/4

elliott-beach committed:
add maxBlockNumber method for fileSystem, and it works
